### PR TITLE
Refactor ThingManagerImpl

### LINKED ISSUE #1924: Update existing Things for definition changes
<a alt="bounty" href="https://www.bountysource.com/issues/95151739"></a>

Currently you have to manually recreate Things for changes in their ThingType or Channel(Group)Definitions to be applied.

The framework should for instance automatically add new channels or configuration parameters.

See also this ESH issue: https://github.com/eclipse-archived/smarthome/issues/2555

These PRs have already been created for solving the issue:

* https://github.com/openhab/openhab-core/pull/1027
* https://github.com/openhab/openhab-core/pull/628
* https://github.com/openhab/openhab-core/pull/527
* https://github.com/eclipse-archived/smarthome/pull/6537

### PR DESCRIPTION
**Pull request by J-N-K**
Closes #1924 

This

- moves config description URI to type-base class
- decouples the thing manager from bundle loading
- makes sure that all thing/channel-types and config descriptions are available when the thing is initialized
- enables thing type updates

Supersedes #3191

Signed-off-by: Jan N. Klug <github@klug.nrw>

### CODE REVIEW

**Comment by jlaur** [comment]
@J-N-K - do you know if this will also fix this issue that I'm seeing from time to time:
```
2023-01-21 09:31:08.326 [WARN ] [core.thing.internal.ThingManagerImpl] - No config description for 'thing-type:energidataservice:service' found when normalizing configuration for 'energidataservice:service:energidataservice'. This is probably a bug.
```

It sounds very much like it, just seeking confirmation.

**Comment by J-N-K** [comment]
Yes.

File: bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/internal/ThingHandlerCallbackImpl.java
Comment (jlaur) [inline bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/internal/ThingHandlerCallbackImpl.java:121]: Good change. Note to self: This will break:
- cbus
- km200
- kostalinverter

as they are setting `ThingStatus.UNINITIALIZED`. They can be fixed beforehand, only five places it seems (from GitHub search, may need to revise).

File: bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/internal/ThingManagerImpl.java
Comment (J-N-K) [inline bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/internal/ThingManagerImpl.java:239]: @jlaur, this code was already there, I just moved it from an anonymous class in `ThingHandlerImpl` to a separate class. If it didn't work, that would be a bug.

Comment (jlaur) [reply bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/internal/ThingManagerImpl.java:239]: You are right. I will test later what the actual current behavior is. I now also noticed this slight inconsistency which seems to have been made on purpose: 6258268031ecc4dedc832f59de01152686fc3cf8. Not sure why this invalid transition would need to be handled "softer" than others.

Comment (jlaur) [reply bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/internal/ThingManagerImpl.java:239]: I finally got around to test this. They are already broken today:
```
java.lang.IllegalArgumentException: Illegal status UNINITIALIZED. Bindings only may set UNKNOWN, ONLINE, OFFLINE or REMOVED.
	at org.openhab.core.thing.internal.ThingManagerImpl$1.ensureValidStatus(ThingManagerImpl.java:230) ~[?:?]
	at org.openhab.core.thing.internal.ThingManagerImpl$1.statusUpdated(ThingManagerImpl.java:193) ~[?:?]
	at org.openhab.core.thing.binding.BaseThingHandler.updateStatus(BaseThingHandler.java:380) ~[?:?]
	at org.openhab.core.thing.binding.BaseThingHandler.updateStatus(BaseThingHandler.java:404) ~[?:?]
```

**Comment by kaikreuzer** [comment]
FTR: I have started reviewing this PR today, but I will need a few more days, since it is a lot of code and it is touching the core mechanisms - but so far all looks good. 👍

File: bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/internal/update/ThingUpdateInstructionReader.java
Comment (kaikreuzer) [reply bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/internal/update/ThingUpdateInstructionReader.java:65]: Another thought, looking at our OH-INF folder: So far, all metadata that binding developers are expected to deliver are XML files (and property files, but those are typically generated).

From a consistency point of view: Might it make sense to also switch to XML for the update information and provide a proper XSD for them as well?

File: bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/internal/ThingManagerImpl.java
Comment (kaikreuzer) [inline bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/internal/ThingManagerImpl.java:1077]: How about sorting them by version?

File: bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/internal/ThingManagerImpl.java
Comment (kaikreuzer) [inline bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/internal/ThingManagerImpl.java:1073]: You will need a different treatment for textually defined things. With the current implementation, they are always considered to be version "0", which means that the logic below is always performed (and the info logging happens on every start again and again).
Instead, for textual things we can assume that they are always of the latest thing-type version, since their channels are typically created on startup based on the latest thing-type definition.

**Review by kaikreuzer — CHANGES_REQUESTED** [review]
Many thanks, @J-N-K, this looks overall pretty good!
Please find my review comments below. I went through the code, but didn't yet do any extensive testing, though - but I guess that can be done once merged, we should just not do it right before releasing a milestone.

File: bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/internal/ThingHandlerCallbackImpl.java
Comment (J-N-K) [reply bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/internal/ThingHandlerCallbackImpl.java:80]: TBH I don't even know what the author wants to say bay this comment (I only copied it when refactoring the inner class to a separate class file). I would say we can omit the word.

File: bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/internal/ThingManagerImpl.java
Comment (J-N-K) [reply bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/internal/ThingManagerImpl.java:131]: I'm not sure. In my original version I did not have this at all, but while testing I realized that the `systeminfo` binding does something really strange: It has a seperate thing-type for each thing with a `ThingTypeUID` `systeminfo:computer-mba-jan-wifi_janessa_me` where the last part is the id-part of the `ThingUID`. The factory reports "supported" for every `ThingTypeUID` that starts with `systeminfo:computer-`. The `ThingType` itself added dynamically to the `ThingTypeRegistry` during the initialization of the thing, so at the time the `ThingManager` checks if everything needed to initialize the thing is present in the registries it fails. IMO this is a faulty behavior of the binding, if it dynamically creates `ThingTypes` it should add those to the registry and make sure that they are present when a factory reports they are supported. WDYT?

File: bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/internal/ThingManagerImpl.java
Comment (J-N-K) [reply bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/internal/ThingManagerImpl.java:1073]: Luckily this is not the case. Since textually defined things are re-created each time the `.things` file is read, they are automatically created with the latest version (given the developer added the correct version property to the thing-type definition).

File: bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/internal/update/UpdateChannelInstructionImpl.java
Comment (J-N-K) [reply bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/internal/update/UpdateChannelInstructionImpl.java:61]: That's a good question. It might be very difficult to get it right here. If the `ChannelType` changes, there is no guarantee that the "new" `ChannelType` requires the same configuration as the old one. Maybe we could apply a policy here: If the configuration shall be copied, use `UPDATE_CHANNEL`, if it shall be discarded, use `REMOVE_CHANNEL` and `ADD_CHANNEL`. WDYT?

File: itests/org.openhab.core.thing.tests/src/main/java/org/openhab/core/thing/internal/update/ThingUpdateOSGiTest.java
Comment (J-N-K) [reply itests/org.openhab.core.thing.tests/src/main/java/org/openhab/core/thing/internal/update/ThingUpdateOSGiTest.java:270]: For `updatedThing` or `oldThing`? IMO Eclipse is wrong here, because `assertThat` allows `null` as actual value and `sameInstance` uses `==` for matching, which is also null-safe. 

Can you check if `assertThat(updatedThing, is(notNullValue()));` solves this issues?

File: bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/internal/ThingManagerImpl.java
Comment (kaikreuzer) [reply bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/internal/ThingManagerImpl.java:131]: Yes, I agree that this sounds like a pretty weird behavior. Did you find out why at all different types are created for all things? That does not seem to make much sense...

File: bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/internal/update/UpdateChannelInstructionImpl.java
Comment (kaikreuzer) [reply bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/internal/update/UpdateChannelInstructionImpl.java:61]: Yes, that would make sense.

File: bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/internal/ThingManagerImpl.java
Comment (kaikreuzer) [reply bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/internal/ThingManagerImpl.java:1073]: > given the developer added the correct version property to the thing-type definition

Why is this necessary at all? Shouldn't we simply assume that the thing-type is always the latest version, which can be derived from the update information? I do not really see a need to additionally specify the version in the thing-type xml - this is rather something that might easily be forgotten or getting out of sync with the update files.

File: itests/org.openhab.core.thing.tests/src/main/java/org/openhab/core/thing/internal/update/ThingUpdateOSGiTest.java
Comment (kaikreuzer) [reply itests/org.openhab.core.thing.tests/src/main/java/org/openhab/core/thing/internal/update/ThingUpdateOSGiTest.java:270]: If you mean adding this line additionally, then the answer is no - the error remains.
But I've just found out that the issue is resolved by declaring `updatedThing` to be nullable:
```
            @Nullable Thing updatedThing = thingRegistry.get(thingUID);
            assertThat(updatedThing, is(not(sameInstance(oldThing))));
```

File: itests/org.openhab.core.thing.tests/src/main/java/org/openhab/core/thing/internal/update/ThingUpdateOSGiTest.java
Comment (kaikreuzer) [reply itests/org.openhab.core.thing.tests/src/main/java/org/openhab/core/thing/internal/update/ThingUpdateOSGiTest.java:278]: Likewise:
```
        @Nullable String thingTypeVersion = updatedThing.getProperties().get(PROPERTY_THING_TYPE_VERSION);
```

File: bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/internal/ThingManagerImpl.java
Comment (mherwege) [reply bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/internal/ThingManagerImpl.java:131]: The reason for this in the systeminfo binding to have thing types by thing is that thing specific channel groups are added. In my understanding, this has to be done on the thing type level. E.g. a system with 5 harddisks will get 5 storage channel groups created dynamically. This cannot be defined upfront across all things.
I am open for alternative suggestions for this if it avoids complexity in core. There are probably other bindings doing this for the same reason though.

Comment (J-N-K) [reply bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/internal/ThingManagerImpl.java:131]: Digitalstrom, Homematic, MQTT, Netatmo and Systeminfo use `ThingTypeProvider`, I did not check in which way. IMO having different thing-types is fine (also your specific case should be handled by dynamic channel addition), my problem is that you report support for a given `ThingTypeUID` but you don't add the `ThingType` to the `ThingTypeRegistry`.

File: bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/internal/ThingManagerImpl.java
Comment (J-N-K) [reply bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/internal/ThingManagerImpl.java:1073]: I added a check and only update if the provider is a `ManagedProvider`

File: bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/internal/update/RemoveChannelInstructionImpl.java
Comment (J-N-K) [reply bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/internal/update/RemoveChannelInstructionImpl.java:40]: The "empty" check is performed before the object ist created, I added a copy.

File: bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/internal/ThingManagerImpl.java
Comment (J-N-K) [reply bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/internal/ThingManagerImpl.java:1077]: I don't think that's a good idea. Especially if we use REMOVE/ADD for updating channels without configuration, we have to preserve the original order (at least for these instructions), otherwise the builder will fail. So it's safer to not mess with the order.

File: bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/internal/update/ThingUpdateInstructionReader.java
Comment (J-N-K) [reply bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/internal/update/ThingUpdateInstructionReader.java:65]: I agree on moving to the OH-INF folder, that avoids future collisions. 

I'm not so sure regarding the XML (although I see the benefit of validating the content with the XSD and having proper documentation). The reason is that parsing XML with XStream is really a PITA. If we can agree on using Jakarta XML, I'm fine with it.

File: bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/internal/update/UpdateChannelInstructionImpl.java
Comment (J-N-K) [reply bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/internal/update/UpdateChannelInstructionImpl.java:30]: That would be a good argument for the XML. The more "options" we habe, the more difficult is it to add it to the simple properties file we have now.

File: bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/internal/update/ThingUpdateInstructionReader.java
Comment (J-N-K) [reply bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/internal/update/ThingUpdateInstructionReader.java]:
```xml
<?xml version="1.0" encoding="UTF-8" standalone="yes" ?>
<update:update-descriptions xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
	xmlns:update="https://openhab.org/schemas/update-description/v1.0.0"
	xsi:schemaLocation="https://openhab.org/schemas/update-description/v1.0.0 https://openhab.org/schemas/update-description-1.0.0.xsd">

	<thing-type uid="testBinding:testThingTypeAdd">
		<instruction-set targetVersion="1">
			<add-channel id="testChannel1">
				<channelTypeUid>testBinding:testChannelTypeId</channelTypeUid>
			</add-channel>
			<add-channel id="testChannel2">
				<channelTypeUid>testBinding:testChannelTypeId</channelTypeUid>
				<label>Test Label</label>
			</add-channel>
			<add-channel id="testChannel3">
				<channelTypeUid>testBinding:testChannelTypeId</channelTypeUid>
				<label>Test Label</label>
				<description>Test Description</description>
			</add-channel>
		</instruction-set>
	</thing-type>

	<thing-type uid="testBinding:testThingTypeRemove">
		<instruction-set targetVersion="1">
			<remove-channel id="testChannel"/>
		</instruction-set>
	</thing-type>

	<thing-type uid="testBinding:testThingTypeUpdate">
		<instruction-set targetVersion="1">
			<update-channel id="testChannel">
				<channelTypeUid>testBinding:testChannelNewTypeId</channelTypeUid>
				<label>New Test Label</label>
			</update-channel>
		</instruction-set>
	</thing-type>

</update:update-descriptions>
```

WDYT?

File: bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/internal/ThingManagerImpl.java
Comment (kaikreuzer) [reply bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/internal/ThingManagerImpl.java:131]: The issue might be that the specific thing types can only be determined/defined as soon as thing handlers become active - and then it feels too late for adding them to the `ThingTypeRegistry`.
But let's not  try to solve this here as part of this PR, but rather live with the implementation you have done. If this causes issues or we have good ideas how to refactor the system info binding, we can any time address this and simplify the logic.

File: bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/internal/update/ThingUpdateInstructionReader.java
Comment (kaikreuzer) [reply bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/internal/update/ThingUpdateInstructionReader.java]: Much more boilerplate, but that's what comes with XML. I like the fact that we can easily check it against XSDs.

I agree that parsing with XStream is no fun, I'll have a look at your Jakarta implementation.

File: bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/internal/ThingManagerImpl.java
Comment (kaikreuzer) [reply bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/internal/ThingManagerImpl.java:1078]: Ok, let me precise my question:
Shouldn't they sorted by `targetVersion`, so that we can be sure that changes for version 2 are never applied before the changes for version 1? People might add new information in the XSD always to the top, we cannot guarantee any order here.
I'm fine with keeping the order of instructions for the same target version, though.

Comment (J-N-K) [reply bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/internal/ThingManagerImpl.java:1078]: Done.

**Review by kaikreuzer — APPROVED** [review]
Wow, you're incredibly fast - thanks!
I think you have addressed all my comments and I am very happy about the result - so let's wait for the build to succeed and we can merge. 👍

**Comment by jlaur** [comment]
Do we need some documentation to go with this change?

**Comment by J-N-K** [comment]
Maybe we can add something here: https://www.openhab.org/docs/developer/bindings/#describing-things

**Comment by lolodomo** [comment]
> * enables thing type updates

Is it documented somewhere what a binding maintainer and/or final user should now do? Or is it handled fully transparently by the core framework?

**Comment by J-N-K** [comment]
@lolodomo You have to tell the framework what to do by creating update instructions. They need to be located in the OH-INF/update directory. The format is described by the XSD (https://github.com/openhab/openhab-core/blob/main/bundles/org.openhab.core.thing/src/main/resources/xsd/update-description-1.0.0.xsd). Examples can be found in `test/resources/OH-INF/update`.

**Comment by lolodomo** [comment]
> @lolodomo You have to tell the framework what to do by creating update instructions. They need to be located in the OH-INF/update directory. The format is described by the XSD (https://github.com/openhab/openhab-core/blob/main/bundles/org.openhab.core.thing/src/main/resources/xsd/update-description-1.0.0.xsd). Examples can be found in `test/resources/OH-INF/update`.

Thank you. This should certainly be documented in our doc for DEV.
This is additional information but the last description/current description of the thing still has to be maintained in regular files in OH-INF/thing ?

**Comment by jlaur** [comment]
> Maybe we can add something here: https://www.openhab.org/docs/developer/bindings/#describing-things

That makes sense. We can always move it later if we find a more suitable space (or create links from relevant places).

**Comment by kaikreuzer** [comment]
FTR: I have uploaded the new XSD with https://github.com/openhab/website/pull/397 to our website.

**Comment by wborn** [comment]
Looks like these changes break the add-ons build, see:

* https://ci.openhab.org/job/openHAB-Addons/985/
* https://github.com/openhab/openhab-addons/actions/runs/4196737262/jobs/7278112750

If I revert these changes locally the tests succeed again. Can you have a look @J-N-K?

**Comment by wborn** [comment]
A more descriptive title would also be nice for the release notes and if it is resolved you can claim the bounty. :wink:

**Comment by J-N-K** [comment]
> Looks like these changes break the add-ons build, see:
> 
> * https://ci.openhab.org/job/openHAB-Addons/985/
> * https://github.com/openhab/openhab-addons/actions/runs/4196737262/jobs/7278112750
> 
> If I revert these changes locally the tests succeed again. Can you have a look @J-N-K?

See https://github.com/openhab/openhab-addons/pull/14431

**Comment by jlaur** [comment]
@J-N-K - I guess I have to be the one asking the stupid question. 😊 I understand from https://github.com/openhab/openhab-addons/pull/14447#issuecomment-1436514658 that update instructions are needed when adding new channels. What is the reason that this cannot be auto-detected? Performance? I can understand more complex scenarios needing instructions, but it seems to me that if a channel doesn't exist in jsondb, but is declared in the binding, then it must be added? I haven't looked into the jsondb data model as I'm personally using files, so there might be a simple explanation.

**Comment by J-N-K** [comment]
There are different reason:

- A missing channel (in the thing, existing in the XML) could have been removed by the thing itself (because it is not available). This is the reverse action of dynamically adding channels (which is more common) but still a valid option. 
- An existing channel (in the thing, but missing in the XML) could have been dynamically added by the thing itself or be an extensible channel that has been added by the user.
- A channel could change (e.g. the channel-type changes from `deconz:switch` to `system:power`, the label and/or description could change).

We have tried for five years to find a solution that autodetects such changes (there have been several attempts here and in ESH), but there have always been issues (which are not absolute corner-cases) where an automatic solution failed. Especially the first two points make it very hard and the solution with the update instruction is far more fail-safe - it requires a deliberate action of the developer to change existing things.

**Comment by jlaur** [comment]
@J-N-K - I see, thanks for the summary/additional context.

**Comment by J-N-K** [comment]
Technically it is ab "option". But IMO there should be policy to upgrade everything that is possible to make a smoother user experience.

**Comment by lolodomo** [comment]
@J-N-K : is there a special syntax to mention the adding or removal of a thing configuration parameter ? Or is it just unnecessary to consider that ?

**Comment by J-N-K** [comment]
It is unnecessary because that depends on the config description and config descriptions are always build when the bundle loads/reloads, so they are automatically updated. For managed things the channels are stored in the database and the thing description is only used when creating the thing and from then on only "restored" from the database (as opposed to the file-base things, there the thing is always "deleted" and "created" when the file is unloaded/loaded).

**Comment by lolodomo** [comment]
Ok.
And if there is a change of the channel type (change of UID), is it necessary to provide something or not ?

**Pull request merged**
